### ISSUE TITLE
Switching to Qt 6 by default

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#7202]
*(Continuation of #5395 because that was getting rather big)*

# Current state

The qt6-v2 branch **has been merged to master** on March 18th, 2023, but with Qt 5 still being the default. See #7628 for details. This issue is now repurposed on what there needs to be done for Qt 6 to be the default Qt wrapper.

~~The last fixes for the `qt6-v2` branch with Qt 6 support went in around May 19th. Multiple core developers (and users) have been using it as a daily driver since far longer (late 2021 / early 2022) and all known crashes are fixed, with all tests passing.~~

~~There is still a lot of remaining work, however, on the development side. Due to the [big amount of changes](https://github.com/qutebrowser/qutebrowser/compare/qt6-v2) (currently 156 commits, 314 files changed, 12,978 additions and 4,094 deletions) it's tricky to integrate this into `master` without breaking lots of existing open PRs. Thus, we'd like to do another round of PR merging before attempting to do so. The current approach also breaks various tooling, by not properly supporting static analysis (#7128, but that will introduce another big code churn), or by not fixing linting issues yet (#1455).~~

~~Bottomline: Feel free to use the `qt6-v2` branch, keep in mind it will be rebased from time to time, and it will still take lots of efforts behind the scenes to get it fully integrated.~~

# How to try

## `mkvenv.py`

- **Take a backup of your qutebrowser data directory** (see `:version`) unless you don't mind losing cookies and such when down-/upgrading. This isn't needed if you use `--temp-basedir` below, but if you forget to do so, your data will get upgraded, and as far as I'm aware, there is no way to read it with Qt 5 after that.
- Clone the git repo, e.g. via `git clone https://github.com/qutebrowser/qutebrowser.git`
- `cd qutebrowser`
- ~~Check out the branch using `git checkout qt6-v2`~~
- Create a new separate environment via `python3 scripts/mkvenv.py --venv-dir .venv-qt6 --pyqt-version 6.3`
- Set `QUTE_QT_WRAPPER=PyQt6` somewhere in the environment
- Run `.venv-qt6/bin/qutebrowser --temp-basedir` and verify via `:version` that you're indeed on Qt 6

## Archlinux

[qutebrowser-qt6-git](https://aur.archlinux.org/packages/qutebrowser-qt6-git) is available in the AUR.

## Windows / macOS

Built as part of the [nightly builds](https://github.com/qutebrowser/qutebrowser/actions/workflows/nightly.yml) on CI. Click on the most recent passing build, then find the `qt6` builds under "Artifacts". Note that you will need to be logged in with a GitHub account to download.

**Note for macOS:** Due to a [PyInstaller issue](https://github.com/pyinstaller/pyinstaller/pull/6903), Chromium's [sandboxing](https://chromium.googlesource.com/chromium/src/+/HEAD/docs/design/sandbox_faq.md) is currently disabled **for macOS builds with Qt 6**. This means that there will be no additional layer of protection in case of Chromium security bugs. Thus, it's advised to not use this build in production. Hopefully, this situation [will be resolved before the final 3.0.0 release](https://github.com/qutebrowser/qutebrowser/issues/7278).

## Flatpak

[Published in the `flathub-beta` repo](https://github.com/qutebrowser/qutebrowser/issues/5395#issuecomment-1129936579):

> Adding the flathub-beta remote and installing from the beta branch can be done through an application manager (Gnome Software, Discover, ...) or via the command-line.
>
> ```
> $ flatpak remote-add --if-not-exists flathub-beta https://flathub.org/beta-repo/flathub-beta.flatpakrepo
> $ flatpak install flathub-beta org.qutebrowser.qutebrowser//beta
> ```

# Remaining issues

- [x] Flickering rectangles with AMD graphics: #7398 (seems to be fixed in Qt 6.4)
- [x] Mouse wheel sometimes stops working for me: #7523 (known upsteam Qt issue)
- [x] favicons don't show for some tabs sometimes ~~(TODO: find reproducer, going to eg https://github.com/qutebrowser/qutebrowser/commits/qt6-test and clicking on one of the commits seems to lose the favicon) *([from @toofar](https://github.com/qutebrowser/qutebrowser/issues/5395#issuecomment-1003259843), haven't checked if it's still an issue)*~~ see https://github.com/qutebrowser/qutebrowser/issues/7202#issuecomment-1142021295
- [x] Wrong MPRIS name: #7431

# TODOs

- [x] Get macOS builds to work
- [x] ~~Look at remaining API changes and new features (comment below)~~ moved to #7756
- [ ] Look at remaining `FIXME:qt6` comments
- [x] Make sure qutebrowser has enough feedback (e.g. logging?) about which Qt wrappers it uses and why
- [x] ~~Maybe investigate storing the Qt wrapper to use in a config file so that we don't unexpectedly switch (and destroy data?) if one is unavailable~~
- [ ] Decide what to do about test coverage (switch to Qt 6? Combine?)
- [x] ~~Probably merge some PRs while waiting for more feedback and to avoid crazy conflicts with the next steps~~ moved to v4
- [x] ~~Integrate #7128 if we didn't find a way to make static analysis work otherwise~~ see #7370 and linked discussion and issue
- [x] ~~Adopt `black` (#1455) so that we pass "normal" linting again and don't have crazy long lines~~ moved to v4
- [x] ~~Look at adopting PySide6 again~~ not a priority at the moment
- [x] Check if we can revert 3c4d51a03cefb010e57c1e85c53e70bb77de8387
- [x] Potentially start cherry-picking some stuff to master
- [ ] Figure out why our way of blocking back/forward mouse button presses broke
- [x] Perhaps introduce a Qt 5 warning to get people to switch over

**Comment by The-Compiler** [comment]
# Remaining interesting changes

- [Merge WebEngineHistory's quick implementation into QWebEngineHistory (I9f9a73bb) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/327625)
- [QWebEnginePage: add new 'loading' property and expose QWebEngineLoadingInfo (I77730307) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/352611)
- [Add API for favicon database (I6096ad9a) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/309633)
- [Add QPageRanges to PDF printing (I0a66c4f1) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/356189)
- [Add QWebEngineNewWindowRequest::openIn(QWebEnginePage *) (I15e72d52) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/359814)
- [Add missing FileSelectionMode enum value (Ie6978223) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/349778)
-[Fix inconsistent number of load signals and their order](https://codereview.qt-project.org/c/qt/qtwebengine/+/340965) (not an API change, but relevant for us)
- [Implement dialogs of File System Access API](https://codereview.qt-project.org/c/qt/qtwebengine/+/353879) (no API change, just potentially interesting)
- [Add basic USB device logging](https://codereview.qt-project.org/c/qt/qtwebengine/+/334655) (same)
- [Suppress error pages also for http errors if they are disabled](https://codereview.qt-project.org/c/qt/qtwebengine/+/324923/) (no API change, but interesting)
- [Disable kAllowContentInitiatedDataUrlNavigations (I9398f54b) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/358628)
- [Widgets (like the command text box) now insert test into the primary selection (if available)](https://bugreports.qt.io/browse/QTBUG-59879) (ignore the fix version, it only made it into 6.0 but didn't get a new ticket)
- [[QTBUG-93848] QNetworkInformation: Behind connectivity portal - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-93848)

# Remaining API changes overview

- `QWebEngineDownloadItem` renamed to `QWebEngineDownloadRequest`
   * More signals we could potentially use?
   * `finished` signal renamed to `isFinishedChanged`
- `QWebEngineHistory` changed, now has a `QWebEngineHistoryModel`?
- QWebEnginePage changes
   * `QWebEnginePage::FileSelectionMode::FileSelectUploadFolder`
   * New `navigationRequested` signal, emitting `QWebEngineNavigationRequest` (new class)
   * New `newWindowRequested` signal, emitting `QWebEngineNewWindowRequest` (new class)
   * New `loadingChanged` signal, emitting `QWebEngineLoadingInfo` (new class)
- `QWebEngineScript` changes
   * `isNull()` removed
   * New `sourceUrl()` and `setSourceUrl()`
- `QWebEngineSettings` changes
   * `globalSettings()` moved (?)
   * New `setParentSettings`

**Comment by The-Compiler** [comment]
@toofar I just found this:

- [[QTBUG-103735] QWebEnginePage::iconChanged() not emitted - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-103735)
- [Do not reset favicon for same document navigations (Iee0c09cd) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/413654)

That seems like a fix for the favicon problem you observed. Not sure if we can add a workaround of some sorts.

**Comment by toofar** [comment]
Yay, another case of waiting for someone else to fix it paying off! It looks like that fix is being picked to the 6.2 and 6.3 branches, so no need for a workaround?

**Comment by donaldguy** [comment]
**TL;DR**: a (if unlikely _the_) macOS build problem is that, using Apple's Python, the virtualenv construction is creating a `.venv/bin/python3` that fails in dylib loading, as the lib is not found successfully relative to the binary's (weird macOS-style) canonical location outside the path

---

Apologies if this is covered already somewhere (and if so I'd suggest some sort of mention should be in docs, if not indeed in the script):

I am taking a run at getting qt6-v2 to build on macOS. Still kinda just started, but insofar as the current latest state I see is 
>PyInstaller doesn't seem to work properly

I wanted to point out a thing that's like ... not extremely obvious (and very ... Apple)

When running `./scripts/mkvenv.py --pyqt-type source`, it SIGABORTs pretty early; a crashdump shows:

```
Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: @executable_path/../Python3
Referenced from: /Users/USER/*/python3
Reason: tried: '/Users/donald/qutebrowser/.venv/bin/../Python3' (no such file), '/usr/local/lib/Python3' (no such file), '/usr/lib/Python3' (no such file)
(terminated at launch; ignore backtrace)
```

This is cause

macOS `/usr/bin/python3` (amongst others) is weird (and fake) as:

1. it is inside the `sealed, local, read-only` APFS `/` mount
2. Apple still wants it to be able to be upgraded and redirected in parallel with other parts of an SDK/toolchain. 

So yeah whereas

```
❯ /usr/bin/python3 -c "import sys; print(sys.executable)"
/Applications/Xcode.app/Contents/Developer/usr/bin/python3
```

even this is a lie, cause the reality is that, for the linking to work out, it seems like the canonical location ends up as:
- (in my immediate case) `/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.8/bin/python3.8` 
- (or if I wanted) `/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/bin/python3.8`
- (or more flexibly) `$(xcode-select -p)/Library/Frameworks/Python3.framework/Versions/Current/bin/python3` (with `Current` and `python3` as symlinks to a sibling directory, so linking is consistent)

---

This is to say, on macOS `python3` (unless pre-empted) is apt to resolve to `/usr/bin/python3`, which is a wrapper program

```
❯ otool -L /usr/bin/python3
/usr/bin/python3:
	/usr/lib/libxcselect.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.120.1)
```

which `exec`s python in such a way that its understood argv0 is the `python3` from the `usr/bin` of the active `/Developer` directory, which in turn is **a symlink** (I'd have thought hard but I'd be wrong) to the real location (of the `python3` which is itself a symlink to the dotted version) within the `.framework`

(And honestly, it may actually be supposed to ignore the disk entirely and hit up the in-memory dylib cache? but it still probably needs the canonical path whether or not that corresponds to a real file)

---

For my part I worked around this by  doing 
```
(cd $(git rev-parse --show-toplevel);
  rm .venv/bin/python{,3}; 
  ln -s $(xcrun -f python3) .venv/bin/; 
  ln -s $(xcrun -f python3) .venv/bin/python)
) && ./scripts/mkvenv.py --keep
```

doing this but with hardlinks recreates the issue; so going from scratch you gotta like

```
python3 -m venv --symlinks .venv
./scripts/mkvenv.py --keep --pyqt-type source
```

which gets me back to good old fashioned qt errors; adding `--qt-version 6.3` is getting me more progress now

**Comment by donaldguy** [comment]
Well, I don't know if any of that turns out to be extremely relevant or not - but I do have a Qt6 6.3 M1 build running!

Here is the initial version info: https://paste.the-compiler.org/view/9ec20cf5

---

This seems to get me back to a working browser!

```shell
# getting to scratch: a thing I surely won't regret!
brew uninstall --ignore-dependencies --force qt || :
brew uninstall --ignore-dependencies --force python
rm -rf /opt/homebrew/lib/python3.* 
rm -rf ~/Library/Caches/pip

brew cleanup
rm -rf ~/Library/Caches/Homebrew/downloads/*

# for bonus points, kill the qt deps with fire:
brew uninstall --force --ignore-dependencies assimp brotli dbus double-conversion freetype hunspell icu4c jasper jpeg libb2 libmng libpng libtiff md4c sqlite webp zstd 
# pcre2 is a dep of git, so it probs can be assumed if brew is working
# maybe also don't wanna kill off glib, just for sanity

brew install python # 3.9.13_1, also brings in sqlite @ 3.38.5
gh repo clone qutebrowser/qutebrowser
cd qutebrowser
git checkout qt6-v2
/opt/homebrew/opt/python@3.9/libexec/bin/python scripts/mkvenv.py --pyqt-version 6.3
./.venv/bin/qutebrowser --temp-basedir
```

This builds like this https://paste.the-compiler.org/view/884b4209 

it did generate `arm64 is not a supported architecture for PyQt binaries on darwin, this will most likely fail.` but also `Successfully installed PyQt6-6.3.0 PyQt6-Qt6-6.3.0 PyQt6-WebEngine-6.3.0 PyQt6-WebEngine-Qt6-6.3.0 PyQt6-sip-13.3.1`

Likewise, `Smoke test failed with status -11. You might find additional information in the debug output above.` (and sometimes(?) a SIGABORT calling back to above?); but also the browser opens and at least basic stuff works

(Probably same-ish version info, but from "clean" build as described: https://paste.the-compiler.org/view/56e3a68d)

**Comment by The-Compiler** [comment]
@donaldguy 

> **EDIT:** Well ... I'm not gonna say this isn't possibly interesting eventually, but I suppose its largely irrelevant right now, as ... something I saw somewhere said 3.8 wasn't good enough for Qt6 anyway(?), and 3.9 out of homebrew is not quite so much of a mess

As in Python 3.8 you mean? That should work fine (anything 3.7 to 3.10 should).

> **TL;DR**: a (if unlikely _the_) macOS build problem is that, using Apple's Python, the virtualenv construction is creating a `.venv/bin/python3` that fails in dylib loading, as the lib is not found successfully relative to the binary's (weird macOS-style) canonical location outside the path

I don't think that's the problem in my case. My `python3` (on macOS 10.14.6 Mojave though) points to my brew-installed one, `/usr/local/bin/python3`. I'm guessing this could be a problem with macOS Monterey (12.3) or newer though, as I think there, Python 2 was removed. Unfortunately, I'm always running the oldest macOS version supported by qutebrowser.

In fact, running `python3 scripts/mkvenv.py --pyqt-version 6.3` gives me a working browser too (other than the failing smoke test, which indeed seems like a segfault on exit of some sorts... sigh).

The trouble starts when trying to package it into a dmg: With [tox](https://tox.wiki) installed, all that should be needed is a `tox -e build-release-qt6 -- --skip-docs` if you want to reproduce (which would be much appreciated!).

Then, when running `open dist/qutebrowser.app` or `dist/qutebrowser.app/Contents/MacOS/qutebrowser`, qutebrowser launches, but the renderer process crashes (you only see a grey page) - with the terminal claiming:

```
dyld: Library not loaded: @rpath/QtOpenGL.framework/Versions/A/QtOpenGL
  Referenced from: /Users/florian/proj/qutebrowser/dist/qutebrowser.app/Contents/MacOS/PyQt6/Qt6/lib/QtWebEngineCore.framework/Helpers/QtWebEngineProcess.app/Contents/MacOS/QtWebEngineProcess
  Reason: image not found
08:27:19 ERROR: Renderer process crashed (status 6)
```

and indeed the only thing in the .app is `Contents/MacOS/PyQt6/QtOpenGL.abi3.so` and `Contents/MacOS/QtOpenGL`, while the virtualenv install has a `.../site-packages/PyQt6/Qt6/lib/QtOpenGLWidgets.framework/Versions/A/QtOpenGLWidgets` indeed.

I think it's probably just a matter of convincing PyInstaller to include it...

**Comment by donaldguy** [comment]
Yeah I was just getting to that bit myself (as you were typing it). Like you said, its kinda just a "someone needs to write the hook" thing... but also it doesn't _not_ seem like the confusion I got into with python 3.8's resolution (vis-a-vis ... weird path dependent linking mismatching argv0's in practice)

---
I do have one new and hopefully useful piece of information for you!

>  (other than the failing smoke test, which indeed seems like a segfault on exit of some sorts... sigh).

this segfault occurs (only) when `QT_DEBUG_PLUGINS` is set!

That is: 

```
 ./.venv/bin/python -c 'from qutebrowser.qt.widgets import QApplication; QApplication([])'
```

exits 0, whereas 

```
QT_DEBUG_PLUGINS=1 ./.venv/bin/python -c 'from qutebrowser.qt.widgets import QApplication; QApplication([])'
```

consistently segfaults (and so, so does `mkvenv.py` as written)

I figure there is a chance this is a ~race condition on the draw of the empty app window, but maybe/probably its just the debug output choking on a null string that's found its way into the plugin lookup path? (messing with `QT_QPA_PLATFORM_PLUGIN_PATH` and `QT_PLUGIN_PATH` didn't fix it tho)

---

And Yeah, because I had exported `QT_DEBUG_PLUGIN=1` at some point, my `scripts/dev/build_release.py` runs were also consistently segfaulting at smoketest - that's why I took so long to (get any .app that I could look at) and get to seeing the OpenGL error message 😅

I did earlier notice that during the PyInstaller run there is a stream of 

```
121 WARNING: Cannot find path ./QtQuick3DParticleEffects.framework/Versions/A/QtQuick3DParticleEffects (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick3D/ParticleEffects/libqtquick3dparticleeffectsplugin.dylib)
123 WARNING: Cannot find path ./Qt3DQuickScene2D.framework/Versions/A/Qt3DQuickScene2D (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene2D/libqtquickscene2dplugin.dylib)
123 WARNING: Cannot find path ./Qt3DRender.framework/Versions/A/Qt3DRender (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene2D/libqtquickscene2dplugin.dylib)
123 WARNING: Cannot find path ./Qt3DCore.framework/Versions/A/Qt3DCore (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene2D/libqtquickscene2dplugin.dylib)
123 WARNING: Cannot find path ./QtConcurrent.framework/Versions/A/QtConcurrent (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene2D/libqtquickscene2dplugin.dylib)
123 WARNING: Cannot find path ./Qt3DQuick.framework/Versions/A/Qt3DQuick (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene2D/libqtquickscene2dplugin.dylib)
130 WARNING: Cannot find path ./QtQmlLocalStorage.framework/Versions/A/QtQmlLocalStorage (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/LocalStorage/libqmllocalstorageplugin.dylib)
143 WARNING: Cannot find path ./QtQmlXmlListModel.framework/Versions/A/QtQmlXmlListModel (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQml/XmlListModel/libqmlxmllistmodelplugin.dylib)
156 WARNING: Cannot find path ./Qt3DAnimation.framework/Versions/A/Qt3DAnimation (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene3D/libqtquickscene3dplugin.dylib)
156 WARNING: Cannot find path ./Qt3DLogic.framework/Versions/A/Qt3DLogic (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene3D/libqtquickscene3dplugin.dylib)
156 WARNING: Cannot find path ./Qt3DInput.framework/Versions/A/Qt3DInput (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene3D/libqtquickscene3dplugin.dylib)
156 WARNING: Cannot find path ./Qt3DRender.framework/Versions/A/Qt3DRender (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene3D/libqtquickscene3dplugin.dylib)
156 WARNING: Cannot find path ./Qt3DCore.framework/Versions/A/Qt3DCore (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene3D/libqtquickscene3dplugin.dylib)
156 WARNING: Cannot find path ./QtConcurrent.framework/Versions/A/QtConcurrent (needed by /Users/donald/qutebrowser/.tox/pyinstaller-64-qt6/lib/python3.9/site-packages/PyQt6/Qt6/qml/QtQuick/Scene3D/libqtquickscene3dplugin.dylib)
```

Most (though definitely not all) of these are to do with the 3D libs, which I note have separate PyQt6-3D (and PyQt63D-Qt6) packages (but no corresponding PyInstaller hook in the upstream). I tried adding them to the requirements to little effect

Def could be a red herring or so, but since there is, apparently, direct linking from PyQt6 (core) to them (and given how they are generally related) - it doesn't seem impossible to me that there is some need of them to resolve this path

(Also more broadly it seems like the `.frameworks` should end up in `qutebrowser.app/Contents/Frameworks`... but oh well)

I also wondered if there is a hidden import amongst the already available hooks that we want in PyQt6.QtOpenGLWidgets?

and like whether any of the Qml stuff needs to come along to satisfy the plugin resolution?


---

I don't really know Qt at all, and python is also not my forte so this is all a lot; hopefully I am being more helpful than annoying

but anyway, I realized the sun is rising here as 04:30 just passed so, uh, I should go to bed now.

Cheers

**Comment by The-Compiler** [comment]
I don't think the missing Qt3D libraries are an issue, as we don't use those parts of QML/QtQuick. In an effort to keep this issue a bit more streamlined than the previous ones, I've split the `QT_DEBUG_PLUGINS` segfault into #7257.

As for the PyInstaller issues, I believe I got to the root cause: https://github.com/pyinstaller/pyinstaller/pull/6892

I seem to get a working DMG that way, with some warnings remaining:

- #7255
- #7256

Still need to clean up things a bit, and then I'll upload it for people to try out.

**Comment by The-Compiler** [comment]
FYI, did another rebase on `master` yesterday - should be all fine now after some (short-lived) trouble I had. A few new tests are failing (#7267), but I don't think that's an actual bug, seems more likely to be an issue with the tests of some sorts.

**Comment by tinywrkb** [comment]
I tried today QtWebEngine `v6.4.0-beta1` with Chromium 98.0.4758.90 built against the KDE 6.3 Flatpak runtime (Qt 6.3.x) and WebRTC screen sharing with PipeWire seems to be broken.  
The log does show that Qt added the `WebRTCPipeWireCapturer`, but nothing else from I can tell from a quick glance, and not a single relevant message went through the D-Bus session.  
Maybe will have to backport some patches. I think I remember there was something around Chromium 99 or 100 related to PipeWire, but I might remember it wrong. I don't remember if it was [this](https://bugs.chromium.org/p/chromium/issues/detail?id=1279574) or something else.  
Anyway, I probably won't look at it before the stable Qt 6.4.0 release.

**Comment by tinywrkb** [comment]
The beta channel of the Flathub app is now (finally) updated to QtWebEngine `6.3.1` with Chromium `94.0.4606.126+`.

**Comment by tinywrkb** [comment]
Flatpak test builds with development releases of QtWebEngine `6.4` (Chromium `98.0.4758.90`) are available [here](https://github.com/flathub/org.qutebrowser.qutebrowser/pull/226).  

I believe that this is the first building QtWebEngine with the app, and not having to pull in the Flatpak QtWebEngine base application.  
Testing building QtWebEngine with the app was on my to-do list, so I used the 6.4-beta1 release as an excuse.  
This has the benefits of having control on the QtWebEngine version, the Chromium revision, and being able to update QtWebEngine quicker.  
The drawback are more complicated and much longer build, and compiling this locally is not an option for me ATM.

The test builds are temporary CI artifacts, and they will need to be generated every week or so.  
The current test build is missing the sandboxing patches that make it possible to sandbox render processes, but the next test build will likely to have them.  

The warning about data directory still applies here when switching between 6.3 and 6.4.

**Comment by angelus2014** [comment]
Hello,

I'm not sure if I should report this here or make a seperate issue. I tried qutebrowser-nightly-2022-07-24-dfde979-qt6-macos yesterday and it seems to run fine without any issues. But I did notice however that the memory usage is a lot bigger than with the current stable version. I'm talking about 3-4 times as much memory is being used for the same pages open.


---
With regards,

Angelo Machils

**Comment by Bonoolu** [comment]
Do we already know the reason why white boxes appear on amd GPUs?
Yesterday, i had this bug on my laptop with integrated Intel graphics. This means this bug is no longer AMD only.
I don't know if this matters, but when I setup qutebrowser on my Intel laptop, i just copied these directories from my AMD machine:
.config/qutebrowser
.local/share/qutebrowser
.cache/qutebrowser

**Comment by rien333** [comment]
Haven't noticed this yet on an old intel GPU, FWIW.

**Comment by Bonoolu** [comment]
Maybe it happened because i copied the .cache directory from an AMD GPU machine. Is this possible?

**Comment by rien333** [comment]
I think starting with the `--temp-basedir` (or for short: `-T`) flag exists exactly to test these kind of scenarios. If the bug is still present if you launch qutebrowser that way, this means it isn't related to your .cache folder/any files you copied over.

**Comment by Bonoolu** [comment]
Sadly, i still don't know how to reproduce this bug.

**Comment by The-Compiler** [comment]
I opened #5395 for that particular issues. If you notice other things being off, please open a new issue. I want to avoid this becoming another long and hard to follow thread with support discussions, I already closed #5395 because it became too hard to follow.

**Comment by cizmazia** [comment]
Would you consider releasing Qutebrowser with the Qt 6 support separately (to avoid delays due to the integration of other 3.0 features)? Qutebrowser is very useful for me as is, however, the stale version of the web engine is becoming an issue. Thanks for your hard work!

**Comment by The-Compiler** [comment]
@cizmazia What do you mean with "release" precisely? There are already binaries of it, and various distributions starting packaging it, so it's unclear to me what you're asking for exactly. See the first post here for details.

**Comment by bodograumann** [comment]
On arch you can install the qt6 version from AUR.
I have been using without major issues for a while now.
Cf. https://aur.archlinux.org/packages/qutebrowser-qt6-git

**Comment by The-Compiler** [comment]
...which is all already explained under "How to try" in the issue text above :wink:

**Comment by cizmazia** [comment]
> @cizmazia What do you mean with "release" precisely? There are already binaries of it, and various distributions starting packaging it, so it's unclear to me what you're asking for exactly. See the first post here for details.

Thanks for your response! 

By "release", I meant a stable release. I did not know how stable the nightly builds are.

I am currently using Qutebrowser on macOS most of the time, which you do not recommend for general use due to the disabled Chromium sandboxing.

**Comment by The-Compiler** [comment]
v3.0.0 *will* be that stable release. I might decide to drop some stuff from the [v3.0.0 milestone](https://github.com/qutebrowser/qutebrowser/milestones/v3.0.0) and reschedule those to v3.1.0/v4.0.0 if something takes too much time. But it's not really as much as it looks like, much of it is either already done in the qt6-v2 branch, or at least in progress, or trivial enough to get it in before the release.

**Comment by bodograumann** [comment]
Can you please add qt6-wayland as an optional dependency to the AUR package?

**Comment by The-Compiler** [comment]
@bodograumann There is nothing specific to either Qt 6 or qutebrowser about that - *any* Qt application needs qt6-wayland to run on Wayland (at least without XWayland). qt6-base already has it as an optional dependency, and judging from its [package page](https://archlinux.org/packages/extra/x86_64/qt6-wayland/), it's not a common thing at all to add it as optional dependency to Qt package (there's a group of related packages which has it, but that seems to be an outlier).

**Comment by The-Compiler** [comment]
Quick pre-announcement: I hopped on a call with @toofar (for the first time ever!), and now we have a battle plan™

In short, I'm planning to merge qt6-v2 tomorrow, and push to master, but with the default still set to PyQt5 (and no auto-switch to PyQt6). Thus, hopefully minimal impact, but making it much easier for us to clean up the last kinks, and for people to try it out.

Going to write a bit of a more in-depth announcement for it tomorrow, the biggest problem is that it'll break some of the open PRs, but that's a necessary evil to clean the path forward.

For packagers, we'll write an easy way out to patch things, so that -qt6 packages still can exist, but can switch to the `master` branch with a trivial (e.g. `sed`) patch to switch the default to Qt 6.

We're also holding off from some further changes that would make things even more difficult (e.g. reformatting via black), this is just the first step to get out of this state where everything is a bit deadlocked and happening in two separate branches.

I'll be away for the rest of the month on Monday, but hoping I can tackle some of the remaining tidbits in April, and then we can first switch the switch over for all -git users, and then finally release 3.0. We also decided to move some stuff originally planned for 3.0 to a new [4.0 milestone](https://github.com/qutebrowser/qutebrowser/issues?q=is%3Aopen+is%3Aissue+milestone%3Av4.0.0), since the top priority once I'm back is finally getting this out of the door.

**Comment by The-Compiler** [comment]
The merge is done and working in the `master-qt6` branch. I'm too tired to finish things and write an announcement today (was busy with #7624), but I'll probably end up doing that tomorrow™!

**Comment by The-Compiler** [comment]
:sparkles: And merged: https://github.com/qutebrowser/qutebrowser/discussions/7628 :sparkles:

**Comment by AckslD** [comment]
I just tried this out by pulling latest master and running
```
python scripts/mkvenv.py --pyqt-version 6.3
```
but it ends with the error:
```
==================== Running Qt smoke test ====================
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/axel/dev/qutebrowser/qutebrowser/qt/widgets.py", line 12, in <module>
    from PyQt5.QtWidgets import *
ModuleNotFoundError: No module named 'PyQt5'

Smoke test failed with status 1. You might find additional information in the debug output above.
```
Am I doing something wrong? Do you alse need pyqt5 installed along pyqt6?

**Comment by The-Compiler** [comment]
@AckslD good catch, thanks! This broke with the recent change to not auto-detect the available library just yet... Should be fixed with dabd8ede8.

**Comment by hpfmn** [comment]
I actually tried to run qutebrowser and it fails:

```
❯ . .venv/bin/activate
❯ qutebrowser --temp-basedir
Fatal error: PyQt5.QtCore is required to run qutebrowser but could not be imported! Maybe it's not installed?

The error encountered was:
No module named 'PyQt5'

Please search for the python3 version of PyQt5.QtCore in your distributions packages, or see https://github.com/qutebrowser/qutebrowser/blob/master/doc/install.asciidoc

If you installed a qutebrowser package for your distribution, please report this as a bug.
```

After running `python scripts/mkvenv.py --pyqt-version 6.4`

**Comment by The-Compiler** [comment]
@hpfmn you need to set `QUTE_QT_WRAPPER=PyQt6` in your environment for it to use Qt 6.

**Comment by rnhmjoj** [comment]
Websites are starting to become unusable: I think it's time to switch, even if the Qt 6 branch is not 100% finished.

**Comment by The-Compiler** [comment]
I'm aware (and the branch doesn't really exist anymore, support is already in master). Unfortunately Qt 6.5.1 had some rather bad regressions, so I'm waiting for 6.5.2 ([scheduled for the 28th](https://wiki.qt.io/Qt_6.5_Release)).

I want to avoid a situation where we flip the switch and people instantly want to go back due to regressions. For one because they'll lose their cookies and such when doing that, but also because they might end up configuring Qt 5 and then never switching back to Qt 6 even if those things are long fixed.

**Comment by The-Compiler** [comment]
Next big step done: #7741 is now merged, which in a nutshell means:

- Useful feedback about what Qt wrapper was chosen and why, everywhere it matters (e.g. `:version` and error dialogs)
- New `--qt-wrapper` argument to qutebrowser which can be used to switch
- New `QUTE_QT_WRAPPER=auto` which enables auto-detection (PyQt6 by default, PyQt5 if that is not available)

Final step will be making that the default, planning to do that when Qt 6.5.2 did land. Maybe I'll find the time to check some of the remaining Qt 6 regressions (personally annoyed by #7222 most) in the meantime...

**Comment by The-Compiler** [comment]
Well, I'm ready for when Qt 6.5.2 is out!

- #7755

**Comment by The-Compiler** [comment]
FYI, I deleted the old `qt6-v2` and `qt6-test` branches to avoid people using old stuff without realizing it's stale.

They're preserved at [qutebrowser/experiments](https://github.com/qutebrowser/experiments) if we ever need them in the future.

**Comment by The-Compiler** [comment]
Still waiting for Qt 6.5.2 to be released. It's been planned for Wednesday, but it looks like things are delayed. I found this:

- [Update submodule refs on '6.5.2' in yocto/meta-qt6 (If7af82fa) · Gerrit Code Review](https://codereview.qt-project.org/c/yocto/meta-qt6/+/488937)
- [Update submodules on '6.5.2 in qt/qt5' (I3d71880c) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qt5/+/488936)

Soooo maybe it'll still happen today? Or on Monday?

**Comment by The-Compiler** [comment]
Well, looks like the domain that's hosted on just did a derp:

![image](https://github.com/qutebrowser/qutebrowser/assets/625793/00fc2976-5cd5-4806-9d69-585e3a31d357)

> As of January 1, 2014 the Internet Corporation for Assigned Names and Numbers (ICANN) will mandate that all ICANN accredited registrars begin verifying the Registrant WHOIS contact information for all new domain registrations and Registrant contact modifications.
> 
> Why this domain has been suspended
> Email address has not been verified.
> This is a new domain registration and the Registrant email address has not been verified.
> 
> or
> The Registrant contact data for this domain was modified but still requires verification.
> Specifically the First Name, Last Name and/or email address have been changed and never verified.

Soooo I'm assuming it'll probably be next week :sweat_smile:

**Comment by DemiMarie** [comment]
I strongly recommend dropping Qt5 support in the very near future for security reasons, as Qt5 WebEngine is behind on patches.

**Comment by The-Compiler** [comment]
> I strongly recommend dropping Qt5 support in the very near future for security reasons, as Qt5 WebEngine is behind on patches.

Can you elaborate? The last security patches for QtWebEngine on Qt 5 [were integrated](https://codereview.qt-project.org/q/project:qt/qtwebengine-chromium+branch:87-based) four days ago.

**Comment by DemiMarie** [comment]
Was not aware of that, sorry.

**Comment by The-Compiler** [comment]
As for Qt 6.5.2, it looks like there was another issue found during releasing:

- [[QTBUG-115150] shadow build failure, webengine - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-115150)
- [[QTBUG-115067] [REG 6.5.1 -> 6.5.2] shadow build failure, webengine - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-115067)
- [Speculative fix for potential race condition in devtools build (Icf342eb2) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/490629)

Welp. Let's hope it still happens this week, as next week I'm at Europython and probably not going to have the time to look at qutebrowser stuff much. Fingers crossed!

**Comment by The-Compiler** [comment]
Finally there's been [some news](https://lists.qt-project.org/pipermail/development/2023-July/044156.html):

> unfortunately, Qt 6.5.2 has been delayed. The final test round for the release candidate is currently ongoing, though. My current estimate is that we can release 6.5.2 at the beginning of next week if there are no critical findings on the remaining tests. I’ll update this also to the Qt’s wiki page – I’m sorry that I didn’t do it already earlier (Jani Heikkinen who’s usually managing Qt 6.5.x releases is currently on summer holiday and I’m substituting for him).

And the [wiki page](https://wiki.qt.io/Qt_6.5_Release) now says 18th (next Tuesday).

**Comment by The-Compiler** [comment]
Qt 6.5.2 [has been released](https://www.qt.io/blog/qt-6.5.2-released-1) finally! Looks like an Archlinux update also is [in progress already](https://gitlab.archlinux.org/archlinux/packaging/packages/qt6-base/-/commit/2496a9424e759e8f863ede30d8a75643e29a998c), so I suppose it'll happen somewhen the next days (perhaps on the weekend at the Europython sprints?) :sparkles:

**Comment by The-Compiler** [comment]
Since Qt 6.5.2 was released finally (with 3 weeks of delay), fixing various annoying issues with Qt 6.5.0 and .1, it's time to finally flip the switch to Qt 6 by default!

So that's what I did at the Europython sprints in Prague :slightly_smiling_face: 

Right now `mkvenv.py` will install the PyQt6 Qt6 packages from the Riverbank Computing development package server (which means you end up downloading it twice). This should change early next week, when those packages get officialy released, but I just couldn't wait any longer :laughing: 

Thanks to @RMPR for some nice pair programming sessions and contributions here at the sprint, fixing some of the last rough edges.

If you use qutebrowser from git, it will now automatically use PyQt6 if that's available - if it is not, it will open a warning page telling you that you should probably ensure that it is.

Only a couple of smaller things now remaining until the v3.0.0 release: https://github.com/qutebrowser/qutebrowser/milestone/49

**Comment by The-Compiler** [comment]
![image](https://github.com/qutebrowser/qutebrowser/assets/625793/dfced6a8-7798-431c-81e5-a559e9484ef7)

soooooooon! Hopefully today!

**Comment by The-Compiler** [comment]
v3.0.0 [was released](https://github.com/qutebrowser/qutebrowser/releases/tag/v3.0.0)!

**Comment by markstos** [comment]
Perhaps this can be unpinned now that Qute 3 / QT 6 are the default.

**Comment by The-Compiler** [comment]
Leaving it pinned for now, until/unless we need space for new pinned issues. While I don't see many people still running Qt 5, I still think it's good to make people aware that there was a big change and they should make sure they're on Qt 6 if they haven't taken care of that yet.
